File: src/algorithms/backtracking/stepping_numbers.py
from typing import List


class Solution:

    def countSteppingNumbers(self, low: int, high: int) -> List[int]:
        lis = []

        def dfs(lo: int, hi: int, step_num: int):
            if hi >= step_num >= lo:
                lis.append(step_num)
            if step_num == 0 or step_num > hi:
                return
            last_d = step_num % 10
            step_num_a = step_num * 10 + (last_d - 1)
            step_num_b = step_num * 10 + (last_d + 1)
            if last_d == 0:
                dfs(lo, hi, step_num_b)
            elif last_d == 9:
                dfs(lo, hi, step_num_a)
            else:
                dfs(lo, hi, step_num_a)
                dfs(lo, hi, step_num_b)

        for i in range(0, 10, 1):
            dfs(low, high, i)
        lis.sort()
        return lis

File: src/algorithms/backtracking/Stepping_Numbers.java
package algorithms.backtracking;

import algorithms.tree.TreeNode;

import java.util.*;

/**
 * @author Yangxiao on 10/5/2019.
 */
public class Stepping_Numbers {
	
	public List<Integer> countSteppingNumbers(int low, int high) {
		List<Integer> res = new ArrayList<>();
		for (int i = 0; i <= 9; i++) {
			dfs(low, high, i, res);
		}
		Collections.sort(res);
		return res;
	}
	
	private void dfs(int low, int hi, long stepNum, List<Integer> res) {
		if (stepNum <= hi && stepNum >= low) {
			res.add((int)stepNum);
		}
		if (stepNum == 0 || stepNum > hi) {
			return;
		}
		long lastD = stepNum % 10;
		long stepNumA = stepNum * 10 + (lastD - 1);
		long stepNumB = stepNum * 10 + (lastD + 1);
		
		if (lastD == 0) {
			dfs(low, hi, stepNumB, res);
		} else if (lastD == 9) {
			dfs(low, hi, stepNumA, res);
		} else {
			dfs(low, hi, stepNumA, res);
			dfs(low, hi, stepNumB, res);
		}
	}
	class solution {
		public List<Integer> countSteppingNumbers2(int low, int high) {
			List<Integer> list = new ArrayList();
			
			for (long i = 0; i <= 9; i++) {
				dfs(low, high, i, list);
			}
			Collections.sort(list);
			return list;
		}
		
		private void dfs(int low, int high, long cur, List<Integer> list) {
			if (cur >= low && cur <= high) list.add((int) cur);
			if (cur == 0 || cur > high) return;
			
			long last = cur % 10, inc = cur * 10 + last + 1, dec = cur * 10 + last - 1;
			
			if (last == 0) {
				dfs(low, high, inc, list);
			} else if (last == 9) {
				dfs(low, high, dec, list);
			} else {
				dfs(low, high, inc, list);
				dfs(low, high, dec, list);
			}
		}
	}
}
